### PR TITLE
Added TLS Client Authentication

### PR DESCRIPTION
**Pull request by takjn**
This pull request supports TLS client authentication.

### CODE REVIEW

File: core.go [hunk 1]
```diff
@@ -237,19 +247,58 @@ func (dc *DeviceContext) DisableClaimMode() error {
...
 	d := &DeviceInfo{}
-	if err := makeRESTCall("GET", fmt.Sprintf("/devices/%d", dc.DeviceID), dc.AuthToken, nil, d); err != nil {
+	if err := makeRESTCall("GET", fmt.Sprintf("/devices/%d", dc.DeviceID), dc, nil, d); err != nil {
 		return nil, err
 	}
 
 	return d, nil
 }
 
+func (dc *DeviceContext) buildConfig() (*tls.Config, error) {
```
Comment (ecnerwalawrence) [inline core.go]: please write a unit test

**Review by ecnerwalawrence — CHANGES_REQUESTED** [review]

File: core.go (hunk 1 above)
Comment (takjn) [reply core.go]: added

**Review by ecnerwalawrence — APPROVED** [review]

File: core.go [hunk 2]
```diff
@@ -119,8 +125,12 @@ type (
 	// If on-demand device provisioning is enabled for your MODE project, you
 	// can call ProvisionDevice to create a new DeviceContext.
```
Comment (ethankan) [inline core.go:126]: Please update the GoDoc on the new TLS-related fields in DeviceContext.

File: rest.go [hunk 3]
```diff
@@ -89,9 +89,23 @@ func makeRESTAuth(authToken string) string {
 	return fmt.Sprintf("ModeCloud %s", authToken)
 }
 
-func makeRESTCall(method string, path string, authToken string, data interface{}, resData interface{}) error {
+func makeRESTCall(method string, path string, dc *DeviceContext, data interface{}, resData interface{}) error {
 	url := makeRESTCallURL(path)
 
+	authToken := ""
+	if dc != nil {
+		authToken = dc.AuthToken
```
Comment (ethankan) [inline rest.go]: No need to set `authToken` if TLSClientAuth is used.

File: rest.go [hunk 4]
```diff
@@ -89,9 +89,23 @@ func makeRESTAuth(authToken string) string {
...
+		if dc.TLSClientAuth {
+			config, err := dc.buildConfig()
```
Comment (ethankan) [inline rest.go]: I prefer not having to call `dc.buildConfig()` every time we make a REST call. We can initialize this when the DeviceContext was first created.

Maybe we should add a function to create DeviceContext for TLS client and properly initialize it.

**Review by ethankan — CHANGES_REQUESTED** [review]

File: core.go (hunk 2 above)
Comment (takjn) [reply core.go:126]: updated.

File: rest.go (hunk 3 above)
Comment (takjn) [reply rest.go]: fixed.

File: rest.go (hunk 4 above)
Comment (takjn) [reply rest.go]: I changed `dc.buildConfig()` to `dc.SetPKCS12ClientCertificate()` to be able to initialize device context when the DeviceContext was first created.

**Comment by takjn** [comment]
Thank you for your comments. I fixed. Please re-review.

**Review by ethankan — APPROVED** [review]

**Pull request merged**